File: code/92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m == n) return head;
        ListNode* prev = nullptr;
        int cnt = 1;
        ListNode* p = head;
        while (cnt < m) {
            prev = p;
            p = p->next;
            cnt++;
        }
        ListNode* start = p;
        ListNode* q = p->next;
        cnt++;
        while (cnt <= n) {
            ListNode* r = q->next;
            q->next = p;
            p = q;
            q = r;
            cnt++;
        }
        start->next = q;
        if (m != 1) prev->next = p;
        else head = p;
        return head;
    }
};

File: code/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
            return head
        def reverse(node, m, n):
            t = node 
            if n > 1:
                if m > 1:
                    node.next, t = reverse(node.next, m - 1, n - 1)
                else:
                    t1, t = reverse(node.next, m, n - 1)
                    node.next = t.next
                    t.next = node
                    node, t = t1, node
            return node, t
        return reverse(head, m, n)[0]
        # if m == 1:
        #     cnt = 1
        #     cur, prev = head, None
        #     while cnt <= n:
        #         cur.next, cur, prev = prev, cur.next, cur
        #         cnt += 1
        #     head.next = cur
        #     head = prev
        # else:
        #     cnt = 1
        #     cur, prev = head, None
        #     while cnt < m:
        #         prev = cur
        #         cur = cur.next
        #         cnt += 1
        #     p = cur
        #     cur = cur.next
        #     cnt += 1
        #     while cnt <= n:
        #         cur.next, cur, p = p, cur.next, cur
        #         cnt += 1
        #     prev.next.next = cur
        #     prev.next = p
        # return head
                
        